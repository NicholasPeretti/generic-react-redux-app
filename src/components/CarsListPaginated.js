import React from 'react'
import PropTypes from 'prop-types'
import CarPropType from '../utils/propTypes/car'
import styled from 'styled-components'
import CarsList from './CarsList'
import CarsListSortBox from './CarsListSortBox'
import PaginationNav from './PaginationNav'
import SubTitleBold from './SubTitleBold'
import SubTitle from './SubTitle'
import Style from '../style'
import Skeleton from 'react-loading-skeleton'

const Row = styled.div`
  display: flex;
  flex-direction: row;
  width: 100%;
`

const Column = styled.div`
  display: flex;
  flex-direction: column;
  width: 100%;
`

class CarsListPaginated extends React.Component {
  componentDidMount () {
    this.props.fetchPage()
  }

  render () {
    const {
      cars,
      fetching,
      fetchPage,
      page,
      hasNext,
      hasPrev,
      totalPages,
      setSort,
      mileageSort
    } = this.props

    return (
      <Column>
        <Row>
          <Column>
            <SubTitleBold>Available cars</SubTitleBold>
            <SubTitle style={{ margin: `${Style.spacing.padding3} 0px` }}>
              {fetching ? (
                <Skeleton width={200}/>
              ) : (
                <span>
                  Showing 1 of {totalPages} page{totalPages > 1 && 's'}
                </span>
              )}
            </SubTitle>
          </Column>
          <Column style={{ width: '40%' }}>
            {!fetching && (
              <CarsListSortBox
                setSort={setSort}
                value={mileageSort}
              />
            )}
          </Column>
        </Row>
        <CarsList cars={cars} fetching={fetching}/>
        <PaginationNav
          page={page}
          totalPages={totalPages}
          hasNext={hasNext}
          hasPrev={hasPrev}
          fetchPage={fetchPage}
        />
      </Column>
    )
  }
}

CarsListPaginated.propTypes = {
  cars: PropTypes.arrayOf(CarPropType).isRequired,
  fetchPage: PropTypes.func.isRequired,
  fetching: PropTypes.bool.isRequired,
  page: PropTypes.number.isRequired,
  totalPages: PropTypes.number.isRequired,
  hasNext: PropTypes.bool.isRequired,
  hasPrev: PropTypes.bool.isRequired,
  mileageSort: PropTypes.number,
  setSort: PropTypes.func.isRequired
}

export default CarsListPaginated
