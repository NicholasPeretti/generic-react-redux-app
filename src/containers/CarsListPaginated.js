import { connect } from 'react-redux'
import CarsListPaginated from '../components/CarsListPaginated'
import { getCars, getPage, isFetching, getResultsCount, hasNext, hasPrev, getTotalPages } from '../redux/ducks/cars'
import { fetchPage } from '../redux/middleware/app/cars'

const mapStateToProps = state => ({
  cars: getCars(state),
  page: getPage(state),
  resultsCount: getResultsCount(state),
  fetching: isFetching(state),
  hasNext: hasNext(state),
  hasPrev: hasPrev(state),
  totalPages: getTotalPages(state)
})

const mapDispatchToProps = dispatch => ({
  fetchPage: page => dispatch(
    fetchPage(page)
  )
})

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(CarsListPaginated)
