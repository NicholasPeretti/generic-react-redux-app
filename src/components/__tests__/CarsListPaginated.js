import React from 'react'
import CarsListPaginated from '../CarsListPaginated'
import PresentationalTest from '../../utils/tests/presentational'
import { shallow } from 'enzyme'
import { CarsList as MOCK_CARS } from '../../utils/tests/data'
import { MemoryRouter } from 'react-router-dom'

const WrappedCarsListPaginated = props => (
  <MemoryRouter>
    <CarsListPaginated {...props} />
  </MemoryRouter>
)

const props = {
  cars: [],
  resultsCount: 0,
  fetchPage: jest.fn(),
  fetching: false,
  hasNext: false,
  hasPrev: false,
  page: 1,
  totalPages: 1,
  setSort: jest.fn()
}

describe('CarsListPaginated', () => {
  PresentationalTest(WrappedCarsListPaginated, props)

  describe('While fetching', () => {
    test('Should match snapshot', () => {
      const wrapper = shallow(<WrappedCarsListPaginated {...props} fetching={true}/>)
      expect(wrapper.html()).toMatchSnapshot()
    })
  })

  describe('Show empty list', () => {
    test('Should match snapshot', () => {
      const wrapper = shallow(<WrappedCarsListPaginated {...props}/>)
      expect(wrapper.html()).toMatchSnapshot()
    })
  })

  describe('Show filled list', () => {
    test('Should match snapshot', () => {
      const wrapper = shallow(
        <WrappedCarsListPaginated
          {...props}
          cars={MOCK_CARS}
          resultsCount={MOCK_CARS.length}
        />
      )
      expect(wrapper.html()).toMatchSnapshot()
    })
  })
})
