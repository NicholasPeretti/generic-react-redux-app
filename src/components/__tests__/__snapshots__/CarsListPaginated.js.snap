// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CarsListPaginated Show empty list Should match snapshot 1`] = `"<div class=\\"sc-gqjmRU jFwtAU\\"><div class=\\"sc-gZMcBi bkpuIb\\"><div class=\\"sc-gqjmRU jFwtAU\\"><h3 class=\\"sc-bdVaJa sc-bwzfXH knfVVl\\">Available cars</h3><h3 style=\\"margin:24px 0px\\" class=\\"sc-bdVaJa fkyhBf\\">Showing 0 of 0 results</h3></div></div><div class=\\"sc-gzVnrw dPdjli\\"></div><nav class=\\"sc-dnqmqq dhWCzT\\"><span class=\\"sc-htpNat lptLpj\\"><a class=\\"sc-htoDjs sc-iwsKbI fVuIeh\\">First</a><a class=\\"sc-htoDjs sc-iwsKbI fVuIeh\\">Previous</a><span style=\\"margin:0px 12px\\">Page 1 of 1</span><a class=\\"sc-htoDjs sc-iwsKbI fVuIeh\\">Next</a><a class=\\"sc-htoDjs sc-iwsKbI fVuIeh\\">Last</a></span></nav></div>"`;

exports[`CarsListPaginated Show filled list Should match snapshot 1`] = `"<div class=\\"sc-gqjmRU jFwtAU\\"><div class=\\"sc-gZMcBi bkpuIb\\"><div class=\\"sc-gqjmRU jFwtAU\\"><h3 class=\\"sc-bdVaJa sc-bwzfXH knfVVl\\">Available cars</h3><h3 style=\\"margin:24px 0px\\" class=\\"sc-bdVaJa fkyhBf\\">Showing 10 of 10 results</h3></div></div><div class=\\"sc-gzVnrw dPdjli\\"><article class=\\"sc-bZQynM hEXood\\"><img src=\\"http://localhost:3001/car.svg\\" class=\\"sc-EHOje qtPkP\\"/><section class=\\"sc-ifAKCX hPHZRJ\\"><h3 class=\\"sc-bdVaJa sc-bwzfXH knfVVl\\">BMW X1</h3><span class=\\"sc-htpNat lptLpj\\">Stock # 10079 - 178994 km - Diesel - black</span><span class=\\"sc-htpNat lptLpj\\"><a class=\\"sc-bxivhb gpMpcx\\" href=\\"/10079\\">View details</a></span></section></article><article class=\\"sc-bZQynM hEXood\\"><img src=\\"http://localhost:3001/car.svg\\" class=\\"sc-EHOje qtPkP\\"/><section class=\\"sc-ifAKCX hPHZRJ\\"><h3 class=\\"sc-bdVaJa sc-bwzfXH knfVVl\\">Skoda Octavia</h3><span class=\\"sc-htpNat lptLpj\\">Stock # 10206 - 142074 km - Petrol - black</span><span class=\\"sc-htpNat lptLpj\\"><a class=\\"sc-bxivhb gpMpcx\\" href=\\"/10206\\">View details</a></span></section></article><article class=\\"sc-bZQynM hEXood\\"><img src=\\"http://localhost:3001/car.svg\\" class=\\"sc-EHOje qtPkP\\"/><section class=\\"sc-ifAKCX hPHZRJ\\"><h3 class=\\"sc-bdVaJa sc-bwzfXH knfVVl\\">Dodge Journey</h3><span class=\\"sc-htpNat lptLpj\\">Stock # 10384 - 125976 km - Petrol - green</span><span class=\\"sc-htpNat lptLpj\\"><a class=\\"sc-bxivhb gpMpcx\\" href=\\"/10384\\">View details</a></span></section></article><article class=\\"sc-bZQynM hEXood\\"><img src=\\"http://localhost:3001/car.svg\\" class=\\"sc-EHOje qtPkP\\"/><section class=\\"sc-ifAKCX hPHZRJ\\"><h3 class=\\"sc-bdVaJa sc-bwzfXH knfVVl\\">Mercedes-Benz CLC-Klasse</h3><span class=\\"sc-htpNat lptLpj\\">Stock # 10556 - 144895 km - Petrol - white</span><span class=\\"sc-htpNat lptLpj\\"><a class=\\"sc-bxivhb gpMpcx\\" href=\\"/10556\\">View details</a></span></section></article><article class=\\"sc-bZQynM hEXood\\"><img src=\\"http://localhost:3001/car.svg\\" class=\\"sc-EHOje qtPkP\\"/><section class=\\"sc-ifAKCX hPHZRJ\\"><h3 class=\\"sc-bdVaJa sc-bwzfXH knfVVl\\">Skoda Felicia</h3><span class=\\"sc-htpNat lptLpj\\">Stock # 10577 - 173054 km - Petrol - blue</span><span class=\\"sc-htpNat lptLpj\\"><a class=\\"sc-bxivhb gpMpcx\\" href=\\"/10577\\">View details</a></span></section></article><article class=\\"sc-bZQynM hEXood\\"><img src=\\"http://localhost:3001/car.svg\\" class=\\"sc-EHOje qtPkP\\"/><section class=\\"sc-ifAKCX hPHZRJ\\"><h3 class=\\"sc-bdVaJa sc-bwzfXH knfVVl\\">Volkswagen Golf I</h3><span class=\\"sc-htpNat lptLpj\\">Stock # 10819 - 179730 km - Petrol - black</span><span class=\\"sc-htpNat lptLpj\\"><a class=\\"sc-bxivhb gpMpcx\\" href=\\"/10819\\">View details</a></span></section></article><article class=\\"sc-bZQynM hEXood\\"><img src=\\"http://localhost:3001/car.svg\\" class=\\"sc-EHOje qtPkP\\"/><section class=\\"sc-ifAKCX hPHZRJ\\"><h3 class=\\"sc-bdVaJa sc-bwzfXH knfVVl\\">Mercedes-Benz GT AMG</h3><span class=\\"sc-htpNat lptLpj\\">Stock # 10908 - 150110 km - Petrol - green</span><span class=\\"sc-htpNat lptLpj\\"><a class=\\"sc-bxivhb gpMpcx\\" href=\\"/10908\\">View details</a></span></section></article><article class=\\"sc-bZQynM hEXood\\"><img src=\\"http://localhost:3001/car.svg\\" class=\\"sc-EHOje qtPkP\\"/><section class=\\"sc-ifAKCX hPHZRJ\\"><h3 class=\\"sc-bdVaJa sc-bwzfXH knfVVl\\">Dodge Avenger</h3><span class=\\"sc-htpNat lptLpj\\">Stock # 11160 - 130487 km - Petrol - yellow</span><span class=\\"sc-htpNat lptLpj\\"><a class=\\"sc-bxivhb gpMpcx\\" href=\\"/11160\\">View details</a></span></section></article><article class=\\"sc-bZQynM hEXood\\"><img src=\\"http://localhost:3001/car.svg\\" class=\\"sc-EHOje qtPkP\\"/><section class=\\"sc-ifAKCX hPHZRJ\\"><h3 class=\\"sc-bdVaJa sc-bwzfXH knfVVl\\">Dodge Avenger</h3><span class=\\"sc-htpNat lptLpj\\">Stock # 11281 - 184601 km - Diesel - white</span><span class=\\"sc-htpNat lptLpj\\"><a class=\\"sc-bxivhb gpMpcx\\" href=\\"/11281\\">View details</a></span></section></article><article class=\\"sc-bZQynM hEXood\\"><img src=\\"http://localhost:3001/car.svg\\" class=\\"sc-EHOje qtPkP\\"/><section class=\\"sc-ifAKCX hPHZRJ\\"><h3 class=\\"sc-bdVaJa sc-bwzfXH knfVVl\\">Tesla Model S</h3><span class=\\"sc-htpNat lptLpj\\">Stock # 11290 - 187797 km - Petrol - green</span><span class=\\"sc-htpNat lptLpj\\"><a class=\\"sc-bxivhb gpMpcx\\" href=\\"/11290\\">View details</a></span></section></article></div><nav class=\\"sc-dnqmqq dhWCzT\\"><span class=\\"sc-htpNat lptLpj\\"><a class=\\"sc-htoDjs sc-iwsKbI fVuIeh\\">First</a><a class=\\"sc-htoDjs sc-iwsKbI fVuIeh\\">Previous</a><span style=\\"margin:0px 12px\\">Page 1 of 1</span><a class=\\"sc-htoDjs sc-iwsKbI fVuIeh\\">Next</a><a class=\\"sc-htoDjs sc-iwsKbI fVuIeh\\">Last</a></span></nav></div>"`;

exports[`CarsListPaginated While fetching Should match snapshot 1`] = `"<h3>Loading...</h3>"`;
